source/client-side-operations-timeout/tests/connection-churn.yml

```diff
@@ -0,0 +1,401 @@
+
+      # Create a failpoint to block first op
+      - name: failPoint
+        object: testRunner
+        arguments:
+          client: *failPointClient
+          failPoint:
+            configureFailPoint: failCommand
+            mode: { times: 1 }
+            data:
+              failCommands: ["insert"]
+              blockConnection: true
+              blockTimeMS: 750
+
+      # Execute op with timeout < block time
+      - name: insertOne
+        object: *collection
+        arguments:
+          timeoutMS: 500
+          document: { _id: 3, x: 1 }
+        expectError:
+          isTimeoutError: true
+
+      # Execute a subsequent operation to complete the read when checking out 
+      # the single available connection.
+      - name: findOne
+        object: *collection
+        arguments:
+          filter: { _id: 1 }
+
+    expectEvents:
+      - client: *client
+        events:
+          - commandFailedEvent:
+              commandName: insert
+      - client: *client
+        eventType: cmap
+        events: [] # Expect no connection closure.
+
+  - description: "write op with unsuccessful pending read"
+    operations:
+      - name: createEntities
+        object: testRunner
+        arguments:
+          entities:
+            - client:
+                id: &client client
+                uriOptions:  
+                  maxPoolSize: 1
+                useMultipleMongoses: false
+                observeEvents:
+                  - commandFailedEvent
+                  - connectionClosedEvent
+            - database:
+                id: &database database
+                client: *client
+                databaseName: *databaseName
+            - collection:
+                id: &collection collection
+                database: *database
+                collectionName: *collectionName
+
+      # Create a failpoint to block first op
+      - name: failPoint
+        object: testRunner
+        arguments:
+          client: *failPointClient
+          failPoint:
+            configureFailPoint: failCommand
+            mode: { times: 2 }
+            data:
+              failCommands: ["insert"]
+              blockConnection: true
+              blockTimeMS: 750
+
+      # Execute op with timeout < block time
+      - name: insertOne
+        object: *collection
+        arguments:
+          timeoutMS: 50
+          document: { _id: 3, x: 1 }
+        expectError:
+          isTimeoutError: true
+
+      # The pending read should fail.
+      - name: insertOne
+        object: *collection
+        arguments:
+          timeoutMS: 50
+          document: { _id: 3, x: 1 }
+        expectError:
+          isTimeoutError: true
+
+    expectEvents:
+      - client: *client
+        events:
+          - commandFailedEvent:
+              commandName: insert
+      - client: *client
+        eventType: cmap
+        events: 
+          - connectionClosedEvent:
+              reason: error
```

Comment:
Should this case result in the connection being closed? I propose the connection remain pending for as many operations as it takes to read the response up to some reasonably small limit like 1-5 seconds.

Reply:
Are you concerned that an operation could take a very long time, say 50s, and we have a connection being checked in and out ~50 times to complete the read? I assume this would also require drivers to keep track of the remaining bytes for each CSOT failure while checking out:

```go
n, err := io.CopyN(io.Discard, conn.nc, int64(size))
if err != nil {
	// If the read times out, record the bytes left to read before exiting.
	nerr := net.Error(nil)
	if l := int32(n); l == 0 && errors.As(err, &nerr) && nerr.Timeout() {
		conn.awaitRemainingBytes = ptrutil.Ptr(l + *conn.awaitRemainingBytes)
	}

	return fmt.Errorf("error discarding %d byte message: %w", size, err)
}
```